Storybook build failing in plugin PRs

## Expected Behavior

The storybook builds should not fail when there are no changes to storybook or it's dependencies.

## Current Behavior

Storybook verification build is failing with `Error: Can't resolve 'babel-loader' in '/home/runner/work/backstage/backstage/packages/storybook'`, see e.g. https://github.com/spotify/backstage/pull/651/checks?check_run_id=627548535

## Possible Solution

Might be an issue with the cache, might be an issue with deps sometimes not being built, or something else.

## Steps to Reproduce

? - Maybe this could be reproduced locally by adding a plugin, clean out the repo from dist and node_modules dirs, and then try to run tasks similar to what we're doing with the GitHub actions.

## Context

This also happened for the tech radar PR, where it was possibly fixed by first adding deps and then removing them. https://github.com/spotify/backstage/pull/628

## Your Environment

Build in GitHub Actions

Also for #681, but the error message seems to be different in this https://github.com/spotify/backstage/runs/634375942

```
ERR! Module not found: Error: Can't resolve '@backstage/core' in '/home/runner/work/backstage/backstage/packages/storybook/.storybook'
(node:3228) UnhandledPromiseRejectionWarning: ModuleNotFoundError: Module not found: Error: Can't resolve '@backstage/core' in '/home/runner/work/backstage/backstage/packages/storybook/.storybook'
```

--- a/packages/storybook/.storybook/main.js
+++ b/packages/storybook/.storybook/main.js
@@ -12,36 +12,44 @@ module.exports = {
     'storybook-dark-mode/register',
   ],
   webpackFinal: async config => {
+    const coreSrc = path.resolve(__dirname, '../../core/src');
+
     config.resolve.alias = {
       ...config.resolve.alias,
+      // Point to dist version of theme and any other packages that might be needed in the future
       '@backstage/theme': path.resolve(__dirname, '../../theme'),
     };
-    config.resolve.modules.push(path.resolve(__dirname, '../../core/src'));
-    config.module.rules.push(
-      {
-        test: /\.(ts|tsx)$/,
-        use: [
-          {
-            loader: require.resolve('ts-loader'),
-            options: {
-              transpileOnly: true,
-            },
-          },
-        ],
-      },
-      {
-        test: /\.(js|jsx)$/,
-        loader: 'babel-loader',
-        options: {
-          presets: ['@babel/preset-react'],
-          plugins: ['@babel/plugin-proposal-class-properties'],
-        },
-      },
-    );
+    config.resolve.modules.push(coreSrc);
+
+    // Remove the default babel-loader for js files, we're using ts-loader instead
+    const [jsLoader] = config.module.rules.splice(0, 1);
+    if (jsLoader.use[0].loader !== 'babel-loader') {
+      throw new Error(
+        `Unexpected loader removed from storybook config, ${jsonLoader.use[0].loader}`,
+      );
+    }
+
     config.resolve.extensions.push('.ts', '.tsx');
 
+    // Use ts-loader for all JS/TS files
+    config.module.rules.push({
+      test: /\.(ts|tsx|mjs|js|jsx)$/,
+      include: [__dirname, coreSrc],
+      exclude: /node_modules/,
+      use: [
+        {
+          loader: require.resolve('ts-loader'),
+          options: {
+            transpileOnly: true,
+          },
+        },
+      ],
+    });
+
     // Disable ProgressPlugin which logs verbose webpack build progress. Warnings and Errors are still logged.
-    config.plugins = config.plugins.filter(({ constructor }) => constructor.name !== "ProgressPlugin")
+    config.plugins = config.plugins.filter(
+      ({ constructor }) => constructor.name !== 'ProgressPlugin',
+    );
 
     return config;
   },

--- a/packages/storybook/.storybook/main.js
+++ b/packages/storybook/.storybook/main.js
@@ -16,6 +16,8 @@ module.exports = {
 
     config.resolve.alias = {
       ...config.resolve.alias,
+      // Resolves imports of @backstage/core inside the storybook config, pointing to src
+      '@backstage/core': coreSrc,
       // Point to dist version of theme and any other packages that might be needed in the future
       '@backstage/theme': path.resolve(__dirname, '../../theme'),
     };

--- a/packages/storybook/tsconfig.json
+++ b/packages/storybook/tsconfig.json
@@ -1,0 +1,5 @@
+{
+  "extends": "../../tsconfig.json",
+  "include": [".storybook/**/*"],
+  "compilerOptions": {}
+}

